File: apps/api/internal/handler/notification.go
package handler

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/maneki/api/internal/model"
	"github.com/maneki/api/internal/repository"
)

// NotificationHandler 通知管理处理器
type NotificationHandler struct {
	notificationRepo *repository.NotificationRepository
}

// NewNotificationHandler 创建通知管理处理器
func NewNotificationHandler(notificationRepo *repository.NotificationRepository) *NotificationHandler {
	return &NotificationHandler{
		notificationRepo: notificationRepo,
	}
}

// GetNotifications 获取通知列表
func (h *NotificationHandler) GetNotifications(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("pageSize", "10"))
	if page < 1 {
		page = 1
	}
	if pageSize < 1 || pageSize > 100 {
		pageSize = 10
	}

	unreadOnly, _ := strconv.ParseBool(c.DefaultQuery("unread", "false"))

	query := repository.ListNotificationQuery{
		Page:       page,
		PageSize:   pageSize,
		UnreadOnly: unreadOnly,
	}

	result, err := h.notificationRepo.List(c.Request.Context(), query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    -1,
			"message": "查询失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":     0,
		"message":  "success",
		"data":     result.Data,
		"total":    result.Total,
		"page":     result.Page,
		"pageSize": result.PageSize,
	})
}

// MarkRead 标记单条已读
func (h *NotificationHandler) MarkRead(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    -1,
			"message": "无效的通知ID",
		})
		return
	}

	if err := h.notificationRepo.MarkRead(c.Request.Context(), uint(id)); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    -1,
			"message": "标记已读失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "已标记为已读",
	})
}

// MarkAllRead 标记全部已读
func (h *NotificationHandler) MarkAllRead(c *gin.Context) {
	if err := h.notificationRepo.MarkAllRead(c.Request.Context()); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    -1,
			"message": "标记全部已读失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "全部已读",
	})
}

// GetNotificationStats 获取通知统计
func (h *NotificationHandler) GetNotificationStats(c *gin.Context) {
	count, err := h.notificationRepo.GetUnreadCount(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    -1,
			"message": "获取统计失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "success",
		"data": gin.H{
			"unread_count": count,
		},
	})
}

// CreateNotificationRequest 创建通知请求
type CreateNotificationRequest struct {
	Title   string `json:"title" binding:"required"`
	Content string `json:"content" binding:"required"`
	Type    string `json:"type" binding:"required"`
}

// CreateNotification 创建通知（供 data-service 调用）
func (h *NotificationHandler) CreateNotification(c *gin.Context) {
	var req CreateNotificationRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"code":    -1,
			"message": "请求参数错误: " + err.Error(),
		})
		return
	}

	notification := &model.Notification{
		Title:   req.Title,
		Content: req.Content,
		Type:    req.Type,
		IsRead:  false,
	}

	if err := h.notificationRepo.Create(c.Request.Context(), notification); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"code":    -1,
			"message": "创建通知失败: " + err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"code":    0,
		"message": "通知创建成功",
		"data":    notification,
	})
}

File: apps/web-admin/src/stores/notification.ts
/**
 * 通知状态管理
 */
import { create } from 'zustand'
import { notificationApi, type NotificationItem } from '../services/notification'

interface NotificationState {
  unreadCount: number
  notifications: NotificationItem[]
  loading: boolean

  fetchStats: () => Promise<void>
  fetchList: (params?: { page?: number; pageSize?: number; unread?: boolean }) => Promise<void>
  markRead: (id: number) => Promise<void>
  markAllRead: () => Promise<void>
}

export const useNotificationStore = create<NotificationState>((set, get) => ({
  unreadCount: 0,
  notifications: [],
  loading: false,

  fetchStats: async () => {
    try {
      const result: any = await notificationApi.getStats()
      if (result.code === 0) {
        set({ unreadCount: result.data?.unread_count || 0 })
      }
    } catch {
      // ignore
    }
  },

  fetchList: async (params = {}) => {
    set({ loading: true })
    try {
      const result: any = await notificationApi.getNotifications(params)
      if (result.code === 0) {
        set({ notifications: result.data || [] })
      }
    } catch {
      // ignore
    } finally {
      set({ loading: false })
    }
  },

  markRead: async (id: number) => {
    try {
      await notificationApi.markRead(id)
      set((state) => ({
        unreadCount: Math.max(0, state.unreadCount - 1),
        notifications: state.notifications.map((n) =>
          n.id === id ? { ...n, is_read: true } : n
        ),
      }))
    } catch {
      // ignore
    }
  },

  markAllRead: async () => {
    try {
      await notificationApi.markAllRead()
      set({
        unreadCount: 0,
        notifications: get().notifications.map((n) => ({ ...n, is_read: true })),
      })
    } catch {
      // ignore
    }
  },
}))
